cnd/src/http_api/routes.rs

```diff
@@ -143,54 +181,99 @@ where
-                    let siren_action = make_siren_action(swap_id, action_name);
-                    acc.with_action(siren_action)
-                }))
+    Ok(())
+}
+
+fn add_events(
+    entity: &mut siren::Entity,
+    alpha_events: LedgerEvents,
+    beta_events: LedgerEvents,
+) -> anyhow::Result<()> {
+    let alpha_state_sub = siren::SubEntity::from_entity(
+        siren::Entity::default()
+            .with_class_member("state")
+            .with_properties(alpha_events)
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
+        &["alpha"],
+    );
+    entity.push_sub_entity(alpha_state_sub);
+
+    let beta_state_sub = siren::SubEntity::from_entity(
+        siren::Entity::default()
+            .with_class_member("state")
+            .with_properties(beta_events)
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
+        &["beta"],
+    );
+    entity.push_sub_entity(beta_state_sub);
+
+    Ok(())
+}
+
+async fn next_available_action<S>(facade: Facade, swap: &S) -> anyhow::Result<Option<ActionName>>
+where
+    S: GetRole
+        + DeployAction
+        + InitAction
+        + FundAction
+        + RedeemAction
+        + RefundAction
+        + Clone
+        + AlphaBlockchain
+        + BetaBlockchain
+        + AlphaAbsoluteExpiry
+        + BetaAbsoluteExpiry,
+{
+    if swap.init_action().is_ok() {
+        return Ok(Some(ActionName::Init));
+    }
+
+    if swap.deploy_action().is_ok() {
+        return Ok(Some(ActionName::Deploy));
+    }
+
+    if swap.fund_action().is_ok() {
+        return Ok(Some(ActionName::Fund));
+    }
+
+    if swap.refund_action().is_ok() {
+        let role = swap.get_role();
+        let (expiry, blockchain_time) = match role {
+            Role::Alice => {
+                let expiry = swap.alpha_absolute_expiry().unwrap();
+                let time = match swap.alpha_blockchain() {
+                    Blockchain::Bitcoin => facade.bitcoin_median_time_past().await?,
+                    Blockchain::Ethereum => facade.ethereum_latest_time().await?,
+                };
+                (expiry, time)
+            }
+            Role::Bob => {
+                let expiry = swap.beta_absolute_expiry().unwrap();
+                let time = match swap.beta_blockchain() {
+                    Blockchain::Bitcoin => facade.bitcoin_median_time_past().await?,
+                    Blockchain::Ethereum => facade.ethereum_latest_time().await?,
+                };
+                (expiry, time)
+            }
+        };
+
+        if expiry < blockchain_time {
+            tracing::debug!("We have decided it's time to refund.");
```

Comment:
You can remove this debug comment now or phrase it better (include the times for example)

Reply:
Cool, will remove.

Reply:
bors is running, leaving this for follow up if you feel strongly about it.